Torch Wrapper Serialize

sampathweb opened the pull request:
Adds unit tests for `load_weights` and `model.save` for `TorchModuleWrapper`.

The tests - `test_load_weights`  and `test_serialize_model` fails when `use_batch_norm=True` i.e., when `torch.nn.BatchNorm` is used in the Torch model.  This is likely the reason for #18403 since `resnet18` has BN layers.

Still investigating the issue and the possible fix.

fchollet commented on keras/utils/torch_utils_test.py:
```diff
@@ -25,24 +34,31 @@ def __init__(self, *args, **kwargs):
+        self.assertEqual(len(model.layers), 2)
+        # Linear - Weights, bias, BN - beta, gamma
+        fc_trainable_count = 2
+        if use_batch_norm:
```
Probably need to override `load_own_weights` to support BatchNorm.

sampathweb replied:
I tried the `store.keys()` in `load_own_variables`, but it doesn't contain the `BatchNorm` Mean and Variance variables since they are not part of `model.parameters`.  So not sure what's the best way to capture or register it since `torch` keeps it separate in the `BatchNorm` layer.  Can you provide your thoughts on the comment here - here - https://github.com/keras-team/keras/issues/18403#issuecomment-1734733059

fchollet commented on keras/utils/torch_utils_test.py:
```diff
@@ -53,4 +69,50 @@ def test_module_autowrapping(self):
+        ref_loss = model.evaluate(x_test, y_test)
+        model.save(temp_filepath)
+
+        new_model = saving.load_model(temp_filepath)
```
I don't understand how this could possibly work in the general case if we don't implement `from_config`/etc.

sampathweb replied:
Added `get_config` to `Classifier` in unit test.

fchollet reviewed (commented):
Thanks for the PR!

nkovela1 reviewed (approved):
LGTM! Thank you for cooperating with me on this, Ramesh. Greatly appreciated!

fchollet commented on keras/utils/torch_utils.py:
```diff
@@ -104,3 +112,36 @@ def _track_module_parameters(self):
+        """
+        state_dict = self.module.state_dict()
+        # Save all layer's state keys
+        store[self.name + "._keys"] = list(state_dict.keys())
```
I don't understand why you need to store the list of keys -- the `store` object is already a dict-like and you should be able to query `.keys()` on it. So just doing 

```python
def save_own_variables(self, store):
    state_dict = self.module.state_dict()
    for key in state_dict.keys():
        store[key] = convert_to_numpy(state_dict[key])
```
should be equivalent

sampathweb replied:
There could be multiple `TorchModuleWrapper` in the `Model`. I wanted to identify all the keys that belonged to the specific `TorchModuleWrapper` Layer, hence wanted to store the `keys` that belong to the layer in `store[self.name + '._keys']`.

fchollet replied:
This is already taken into by the serialization logic. Just using keys that are local to the layer is enough.

sampathweb replied:
`state_dict` includes keys for `BatchNorm` `mean` and `variance` that are not part of `model.parameters`, and hence missed by default save logic of layer variables.  So the logic was intended to capture all of the `state_dict` keys.

sampathweb replied:
@fchollet - Thanks for your patience. I didn't have to save the `_keys` or force a `name`.  It works with just storing and loading from `state_dict`.  Tested with multiple wrappers and batch norm.

fchollet commented on keras/utils/torch_utils.py:
```diff
@@ -69,7 +71,13 @@ def call(self, inputs):
-    def __init__(self, module, name=None):
+    def __init__(self, module, name="torch_module_wrapper"):
+        if name is None:
+            raise ValueError(
```
`name` will be autogenerated and should not be required

fchollet replied:
Please add a test with two or three `TorchModuleWrapper`'s. The name special casing is not needed.

sampathweb replied:
Added unit tests for multiple `TorchModuleWrapper` and also for tests for `autowrapping`.  The issue is, each of these models are independent `torch` models and have the same names for their sate variables.  This was the issue the name prefix was added to address. - 
```
(Pdb) model.torch_wrappers
[<TorchModuleWrapper name=torch_module_wrapper_6, built=True>, <TorchModuleWrapper name=torch_module_wrapper_7, built=True>]
(Pdb) model.torch_wrappers[0].state_dict()
OrderedDict([('module.0.weight', tensor([[ 0.4827, -0.3522],
        [-0.4961,  0.3488]])), ('module.0.bias', tensor([-0.1276, -0.0900]))])
(Pdb) model.torch_wrappers[1].state_dict()
OrderedDict([('module.0.weight', tensor([[-0.1081, -0.2072],
        [ 0.2882, -0.2535]])), ('module.0.bias', tensor([ 0.5350, -0.5053]))])
```

sampathweb replied:
Issue fixed.  I didn't have to save `_keys` or enforce names to be same.  It works well because `store` only gives its keys to the layer.  I didn't realize that.  I thought `store` gives ALL the keys of the model so there could be clashes.  Thanks for your patience in this review.

fchollet reviewed (commented):
Thanks for the update!

Question: what happens if we do

```
mw = TorchModuleWrapper(module)
config = mw.get_config()
TorchModuleWrapper.from_config(config)
```

?

I suspect we might need to override from_config/get_config to use `pickle` or something along those lines.

sampathweb commented:
> Thanks for the update!
> 
> Question: what happens if we do
> 
> ```
> mw = TorchModuleWrapper(module)
> config = mw.get_config()
> TorchModuleWrapper.from_config(config)
> ```
> 
> ?
> 
> I suspect we might need to override from_config/get_config to use `pickle` or something along those lines.

Right.  
- Added test for `test_from_config`.
- Added `from_config` and `get_config` to use `torch.load` and `torch.save` to serialize to bytes. Torch customizes the pickle process so used it rather than directly using pickle.

sampathweb commented:
The TF Error is not relevant to this CL.  Will update GitHub Action to be robust to `codecov` service not being available.

```
[2023-10-06T16:12:22.463Z] ['info'] Pinging Codecov: https://codecov.io/upload/v4?package=github-action-3.1.4-uploader-0.6.2&token=*******&branch=torch-wrapper-serialize&build=6433876874&build_url=https%3A%2F%2Fgithub.com%2Fkeras-team%2Fkeras%2Factions%2Fruns%2F6433876874&commit=3e62eed1cdfa59b51a4fbe48034664078500e291&job=Tests&pr=18498&service=github-actions&slug=keras-team%2Fkeras&name=&tag=&flags=keras%2Ckeras-tensorflow&parent=
[2023-10-06T16:12:52.517Z] ['error'] There was an error running the uploader: Error uploading to [https://codecov.io:](https://codecov.io/) Error: There was an error fetching the storage URL during POST: 502 - 
<html><head>
<meta http-equiv="content-type" content="text/html;charset=utf-8">
<title>502 Server Error</title>
</head>
<body text=#000000 bgcolor=#ffffff>
<h1>Error: Server Error</h1>
<h2>The server encountered a temporary error and could not complete your request.<p>Please try again in 30 seconds.</h2>
<h2></h2>
</body></html>

Error: Codecov: Failed to properly upload: The process '/home/runner/work/_actions/codecov/codecov-action/v3/dist/codecov' failed with exit code 255
```

fchollet commented on keras/utils/torch_utils.py:
```diff
@@ -69,8 +73,8 @@ def call(self, inputs):
 
-    def __init__(self, module, name=None):
-        super().__init__(name=name)
+    def __init__(self, module, name=None, *args, **kwargs):
```
You can remove `args` here and below

fchollet reviewed (commented):
Looks good, thanks for the update!

fchollet reviewed (approved):
LGTM

martin-gorner commented:
It sounds like a bad idea to silently use pickle in get_config() / from_config(). Users assume get_config() / from_config() to be safe because thy are limited to loading data only. If they suddenly start loading pickled code, the security model changes but users might not be aware of it.
